umpost/um2netcdf.py

```diff
@@ -319,6 +321,66 @@ def fix_latlon_coords(cube, grid_type, dlat, dlon):
     fix_lon_coord_name(longitude_coordinate, grid_type, dlon)
 
 
+def get_default_fill_value(cube):
+    """
+    Get the default fill value for a cube's data's type.
+
+    Parameters
+    ----------
+    cube: Iris cube object.
+
+    Returns
+    -------
+    fill_value: numpy scalar with type matching cube.data
+    """
+    if cube.data.dtype.kind == 'f':
+        fill_value = DEFAULT_FILL_VAL_FLOAT
+
+    else:
+        fill_value = default_fillvals[
+            f"{cube.data.dtype.kind:s}{cube.data.dtype.itemsize:1d}"
+        ]
+
+    # NB: the `_FillValue` attribute appears to be converted to match the
+    # cube data's type externally (likely in the netCDF4 library). It's not
+    # strictly necessary to do the conversion here. However, we need
+    # the separate `missing_value` attribute to have the correct type
+    # and so it's cleaner to set the type for both here.
+
+    # TODO: Is there a cleaner way do do the following conversion?
+    return np.array([fill_value], dtype=cube.data.dtype)[0]
```

Comment:
A cleaner way might be possible with numpy's type "constructors":

```python
>>> import numpy as np
>>> x = np.int32(5)  # x is a numpy int
>>> type(x)
numpy.int32
```

This would need a lookup mapping a dtype key to a constructor function, like:
```python
lk = {"int32": np.int32, ...}
# other code

return lk[str(fill_value.dtype)](fill_value)
```
It's a bit more code, but doesn't involve creating a `np` array to convert a single primitive value.

Reply:
We've opted to leave the code as is, as other approaches require us to know all the data types that could appear in a cube. I've added a comment in 656d699 noting that any cleaner approaches should be implemented if they're found, and noted that a similar approach is currently used in the [`netCDF4` package](https://github.com/Unidata/netcdf4-python/blob/188cc6c03a8eef0add64c544667b14a0a7b3c85d/src/netCDF4/_netCDF4.pyx#L4413)